### ISSUE TITLE
Optimize signature verification

### ISSUE THREAD

**Issue by NikVolf** [paritytech/substrate#4908]
1. During block import, they can be split in advance and verified in parallel.
2. During block construction, the fact that transaction previously was when imported in transaction pool can be cached.

(sometime maybe) 3. Some networks that do only something like staking might end up with a lot of signatures from the same private key. Those can be batch-verified.

**Comment by tomusdrw** [comment]
I think `sr` also supports batch-verification of multiple signatures (which is faster than verifying them one-by-one), might be a good idea to incorporate that too. I don't know the details though and if it's already implemented in schnorrkel already, but for sure @burdges will know both :)

**Comment by burdges** [comment]
You can even batch verify with a futures API like https://www.zfnd.org/blog/futures-batch-verification/

There are `verify_batch` fns for both [ed25519](https://docs.rs/ed25519-dalek/1.0.0-pre.1/ed25519_dalek/fn.verify_batch.html) and [schnorrkel](https://docs.rs/schnorrkel/0.8.5/schnorrkel/sign/fn.verify_batch.html).  Also, I made sure calls to [`schnorrkel::Signature::from_bytes`](https://docs.rs/schnorrkel/0.8.5/schnorrkel/sign/struct.Signature.html#method.from_bytes) always fail on ed25519 signatures, so one can write a `verify_batch` for `AnySignature` by assigning the signatures that deserialize as schnorkel to the schnorkel side, and the others to the ed25519 side, and then calls the appropriate `verify_batch` for each, like
```
pub struct AnySignedBatch<'m> {
    sr_t: Vec<merlin::Transcript>,
    sr_sigs: Vec<schnorrkel::Signature>,
    sr_pks: Vec<schnorkel::PublicKey>,
    ed_m: Vec<Cow<'m,[u8]>>,
    ed_sigs: Vec<ed25519_dalek::Signature>,
    ed_pks: Vec<ed25519_dalek::PublicKey>,
}

impl<'m> AnySignedBatch<'m> {
    pub fn add(&mut self, msg: &[u8], signature: [u8; 64], publickey: [u8; 32]) -> Result<()> {
        if Ok(sr_sig) = ::schnorrkel::Signature::from_bytes(signature) {
              let sr_pk = ::schnorrkel::PublicKey::from_bytes(publickey) ?;
              // Do substrate's thing for transcripts
              // Push them on the first three Vecs
        } else if Ok(ed_sig) = ::ed25519_dalek::Signature::from_bytes(signature) {
              let ed_pk = ::ed25519_dalek::PublicKey::from_bytes(publickey) ?;
              // Push them on the second three Vecs
        }
    }

    pub fn verify(self) -> Result<()> {
        ::schnorrkel::sign::verify_batch(
            sr_t.as_slice(),
            sr_sigs.as_slice(),
            sr_pks.as_slice(),
        ) ?;
        ::ed25519_dalek::verify_batch(
            ed_m.iter().chain(|x| x.deref()).collect::<Vec<&[u8]>>().as_slice(),
            ed_sigs.as_slice(),
            ed_pks.as_slice(),
        )
    }
}
```

I'd meant to send a pull request to `AnySignature` for this many months ago, but never did.  I just asked @DemiMarie-parity to do so.

**Comment by bkchr** [comment]
The client side implementation shouldn't be that hard, especially as we use `MultiSignature` which is strongly typed. Integrated this into the runtime is the "much" harder part.

**Comment by NikVolf** [comment]
> You can even batch verify with a futures API like https://www.zfnd.org/blog/futures-batch-verification/

This hardly apply, we don't need to defer something this with futures
We have all transactions, we just run all signature verifications before actual block import (batching or not).

Let alone block import in our case cannot be async, since we can't stop-pause runtime (where block execution actually occurs) afaik

**Comment by bkchr** [comment]
`Future` is here just a programming paradigm used to express the validation. You are right that we can not give back the context, however we can just "block" (e.g. the same that is done when calling the code directly). 

The idea would be that you have your verifiers and push the signatures. Verifiers that do not support batching, can start verifying directly in some background thread. After you have pushed all signatures, you just call "wait()". The verifier that does not supports batching will wait for all signatures to be verified and the ones that support batching will now call the `verify_batched` method.

**Comment by burdges** [comment]
One caveat, these `verify_batch` methods are randomized algorithms that need real randomness.  We do need them to be run as native code though so that the runtime cannot access their source of randomness.  We want most signature verification to run native for performance anyways of course.  

We can safely use them in block verification because the change a valid signature set fails is zero and the chance an invalid signature set passes is smaller than 2^{-128}.  In fact, the false validity rate is so low that they cannot be used by malicious runtimes as a randomness oracle.

I think past benchmarks have shown savings of like 40%, maybe dalek maybe upgraded their multi-scalar multiplication since then.

**Comment by DemiMarie-parity** [comment]
@burdges could we use a deterministic hash of all inputs to provide the randomness?  That is a well-known trick used in deterministic ECDSA.

**Comment by bkchr** [comment]
Why? We do the verification anyway on native.

**Comment by burdges** [comment]
I do believe that's secure, but it's a Fiat-Shamir transform that requires writing down a security proof.  We may do that anyways, but verification should run on native anyways, so it's low priority, and no reason to block this and other work.

**Comment by DemiMarie-parity** [comment]
Another caveat: `verify_batch` will fail if *any* of the signatures are bad.  Does Substrate need to know *which* signature is bad, or do we just reject the entire block?

**Comment by burdges** [comment]
We discussed this above, but not so clearly.  If any signature is bad then we reject the whole block, and thus do not care which one fails.  

We already check signatures when importing transactions into the transaction pool, so block production could use batch verification too, except just caching or trusting those signatures is faster, and this is mostly a question of the interface we want to put in front of parachain devs.

**Comment by bkchr** [comment]
The current implementation just skips verification of signatures for block production. It will depend on the developer, which way they want to choose. Batch verification for block production will not work, because we don't know how many transaction we will be able to include in the block. 

I think skipping verification on block production from inside the runtime is fine, the runtime needs to trust the native side anyway. However, you as a validator don't want to run a native node that you don't trust and could approve signatures that are incorrect.

**Comment by DemiMarie-parity** [comment]
Do we want `merlin` in the runtime?  I am not so sure.  @burdges should the batching be done on the native side instead?

**Comment by NikVolf** [comment]
> Do we want `merlin` in the runtime? I am not so sure. @burdges should the batching be done on the native side instead?

Nothing will be verified at runtime

**Comment by DemiMarie-parity** [comment]
For clarification:
- Is `AnySignedBatch` a runtime type, or should it be disabled when compiling for wasm?
- Can we fit all of the signatures in memory?
- I see no uses of `AnySignature` in the runtime.  Is this one of the tasks?

**Comment by bkchr** [comment]
@DemiMarie-parity @NikVolf already started working on this and the block production part is already finished. The block import batch verification will require some more work and isn't only about `AnySignature`.

**Comment by DemiMarie-parity** [comment]
@bkchr is @NikVolf going to do the verification part as well?

**Comment by NikVolf** [comment]
> @bkchr is @NikVolf going to do the verification part as well?

Yep, no worries, it will be just extension to crypto runtime api. Runtime will deconstruct transactions and pass signatures to this updated api. Little else required on the runtime side as I understand. What @burdges is writing about happens on native side.

**Comment by DemiMarie-parity** [comment]
Ah, okay!  So that is a different problem.
